[16.0][IMP] event_sale_session: Recovery unconfirmed quantity event

carolinafernandez-tecnativa opened the pull request:
- Recovery functionality unconfirmed qty on event.session from v14

@Tecnativa
TT49353

@chienandalu @pedrobaeza

carolinafernandez-tecnativa committed:
[IMP] event_sale_session: Recovery unconfirmed quantity event

carolinafernandez-tecnativa committed:
[IMP] event_sale_session: Recovery unconfirmed quantity event

chienandalu commented on event_sale_session/models/event_event.py:
```diff
@@ -0,0 +1,44 @@
+class EventEvent(models.Model):
+    _inherit = "event.event"
+
+    order_line_ids = fields.One2many(
+        comodel_name="sale.order.line",
+        inverse_name="event_id",
+        string="Sales Order Lines",
+    )
```
There's already a `sale_order_lines_ids` field provided by `event_sale`

chienandalu commented on event_sale_session/models/event_event.py:
```diff
@@ -0,0 +1,44 @@
+        inverse_name="event_id",
+        string="Sales Order Lines",
+    )
+    unconfirmed_qty = fields.Integer(
+        compute="_compute_unconfirmed_qty",
+        store=True,
+    )
+
+    @api.depends(
+        "order_line_ids",
+        "order_line_ids.product_uom_qty",
+        "order_line_ids.order_id.state",
+    )
+    def _compute_unconfirmed_qty(self):
+        for event in self:
+            event.unconfirmed_qty = int(
+                sum(
+                    event.order_line_ids.filtered(
+                        lambda x: x.order_id.state in ("draft", "sent")
+                    ).mapped("product_uom_qty")
+                )
+            )
+
+    def button_open_unconfirmed_event_order(self):
+        action = self.env["ir.actions.act_window"]._for_xml_id("sale.action_quotations")
+        sales = (
+            self.env["sale.order.line"]
+            .search([("event_id", "in", self.ids)])
+            .mapped("order_id")
+            .ids
+        )
+        action["domain"] = [("id", "in", sales), ("state", "in", ("draft", "sent"))]
+        action["context"] = {}
+        return action
```
This stuff doesn't belong to this module (I know it was, but it wasn't right).

carolinafernandez-tecnativa replied:
Our customer needs the information to check from event form, that's why is necessary.

chienandalu replied:
I know: but that doesn't mean that this code belongs here :smile It should either go to its own module or to custom code

ivantodorovich replied:
I agree this "unconfirmed quantity" feature doesn't belong here. This module's purpose should be to implement the event sessions with at most the same features you find in standard Events module, not more.

Any other extra feature should be moved to a separate module. (e.g: `event_sale_session_unconfirmed_qty`? notice you could even have a `event_sale_unconfimed_qty` module that implements this in standard events, because the feature doesn't exist in standard)

chienandalu commented on event_sale_session/tests/test_event_sale_session.py:
```diff
@@ -54,22 +57,6 @@ def test_sale_session(self):
         orders = self.env["sale.order"].search(action["domain"])
         self.assertIn(self.order, orders)
 
-    def test_sale_order_line_session_onchange_autocomplete(self):
-        """Test that session is automatically filled and or unset on form"""
-        form = Form(self.order)
-        line = form.order_line.new()
-        line.product_id = self.product
-        # Case 1: The event is a session event, but has multiple sessions
-        line.event_id = self.event
-        self.assertFalse(line.event_session_id)
-        # Case 2: The event is a session event with only 1 session
-        (self.event.session_ids - self.session).active = False
-        line.event_id = self.event
-        self.assertEqual(line.event_session_id, self.session)
-        # Case 3: The event is not a session event, session should be unset
-        line.event_id = self.env.ref("event.event_0")
-        self.assertFalse(line.event_session_id)
-
```
Why removing this whole test?

carolinafernandez-tecnativa replied:
Onchange is only applied when you are in form view

ivantodorovich replied:
Sorry, I still don't understand the reason to remove it 😓 are you facing an issue with this test while implementing the new feature?

pedrobaeza replied:
Adding this uncovers a problem in a computed method, that has been fixed in an additional commit 9799cfb, so the test is restored.

ivantodorovich reviewed (changes requested):
This should be in a separate `event_sale_session_unconfirmed_qty` module IMO

carolinafernandez-tecnativa commented:
@pedrobaeza what do you think to move this functionality to new module?
Also in v14 unconfirmed qty also was added to event.event and there was a button to check all sale orders with state in draft or sent associated to each event.
Maybe the new module should be called event_sale_unconfirmed_qty so we can add both functionalities, what do you think @ivantodorovich ?

ivantodorovich commented:
> Maybe the new module should be called event_sale_unconfirmed_qty so we can add both functionalities, what do you think @ivantodorovich ?

If it depends on `event_session` then it should be prefixed with it.

The idea to split them in `event_sale_unconfirmed_qty` and `event_sale_session_unconfirmed_qty` is to offer the same feature to people using the event modules but not the event_session module. But if you don't need the feature on the standard events then there's no point in developing it now, still.. it's nice to keep this in mind to structure the modules.

Take for example these two modules:
- event_registration_multi_qty
- event_session_registration_multi_qty

chienandalu commented:
@pedrobaeza please add "no stale" label

pedrobaeza committed:
[FIX] event_sale_session: Correct session from event computation

It was not taking into account if there's no session set before.

It's also avoid side effects it `active_test=False` context comes from
any part of the code.

pedrobaeza reviewed (approved):
@ivantodorovich as the requested change has been fixed, I'm merging this for continuing with the fw-port, as we are even going now to 18.

Thank you all for your comments.

/ocabot merge minor

pedrobaeza commented:
/ocabot merge minor

pedrobaeza commented:
/ocabot merge minor
